File: app/agents/router.py
import base64

from app.agents.prompts import AGENT_PROMPTS, SCENE_TO_AGENT
from app.schemas.insight import AgentId, SceneClassification, SceneType, StructuredInsight
from app.services.vlm import vlm_service


class SceneRouter:
    async def route(
        self,
        image_bytes: bytes,
        agent_override: AgentId | None = None,
        image_caption: str | None = None,
    ) -> AgentId:
        if agent_override:
            return agent_override

        image_b64 = base64.b64encode(image_bytes).decode("utf-8")
        raw = await vlm_service.classify_scene(image_b64, image_caption=image_caption)

        try:
            classification = SceneClassification(
                scene_type=SceneType(raw.get("scene_type", "general")),
                text_density=raw.get("text_density", "none"),
                has_person=bool(raw.get("has_person", False)),
                recommended_agent=AgentId(raw.get("recommended_agent", "general_curiosity")),
                reasoning=raw.get("reasoning", ""),
            )
            return classification.recommended_agent
        except (ValueError, KeyError):
            scene = raw.get("scene_type", "general")
            return SCENE_TO_AGENT.get(scene, AgentId.GENERAL_CURIOSITY)


class InsightPlanner:
    async def analyze(
        self,
        image_bytes: bytes,
        agent_id: AgentId,
        locale: str = "zh-CN",
        image_caption: str | None = None,
        latitude: float | None = None,
        longitude: float | None = None,
    ) -> StructuredInsight:
        image_b64 = base64.b64encode(image_bytes).decode("utf-8")
        system_prompt = AGENT_PROMPTS[agent_id]
        raw = await vlm_service.analyze_image(
            image_b64,
            system_prompt,
            locale=locale,
            image_caption=image_caption,
            latitude=latitude,
            longitude=longitude,
        )
        raw["agent_id"] = agent_id.value
        return StructuredInsight.model_validate(raw)


scene_router = SceneRouter()
insight_planner = InsightPlanner()

File: src/agents/router.ts
import { AGENT_PROMPTS, SCENE_TO_AGENT } from "./prompts.js";
import { sanitizePalmInsight } from "./palmSanitize.js";
import {
  AgentId,
  agentIdSchema,
  normalizePalmPath,
  sceneClassificationSchema,
  structuredInsightSchema,
  type StructuredInsight,
} from "../schemas/insight.js";
import { vlmService } from "../services/vlm.js";

/** 校验前把 palm_lines.path 各种畸形格式归一成 {x,y}[] */
function normalizePalmReadingRaw(raw: Record<string, unknown>): void {
  const reading = raw.palm_reading;
  if (!reading || typeof reading !== "object") return;
  const pr = reading as Record<string, unknown>;
  const lines = pr.palm_lines;
  if (!Array.isArray(lines)) return;
  for (const line of lines) {
    if (!line || typeof line !== "object") continue;
    const l = line as Record<string, unknown>;
    l.path = normalizePalmPath(l.path);
  }
}

export class SceneRouter {
  async route(input: {
    imageBytes: Buffer;
    agentOverride?: AgentId | null;
    imageCaption?: string | null;
  }): Promise<AgentId> {
    if (input.agentOverride) {
      return input.agentOverride;
    }

    const imageB64 = input.imageBytes.toString("base64");
    const raw = await vlmService.classifyScene({
      imageB64,
      imageCaption: input.imageCaption,
    });

    const parsed = sceneClassificationSchema.safeParse({
      scene_type: raw.scene_type ?? "general",
      text_density: raw.text_density ?? "none",
      has_person: Boolean(raw.has_person),
      recommended_agent: raw.recommended_agent ?? "general_curiosity",
      reasoning: raw.reasoning ?? "",
    });

    if (parsed.success) {
      return parsed.data.recommended_agent;
    }

    const scene = String(raw.scene_type ?? "general");
    return SCENE_TO_AGENT[scene] ?? AgentId.GENERAL_CURIOSITY;
  }
}

export class InsightPlanner {
  async analyze(input: {
    imageBytes: Buffer;
    agentId: AgentId;
    locale?: string;
    imageCaption?: string | null;
    latitude?: number | null;
    longitude?: number | null;
    birthday?: string | null;
    /** 手动选择的专项镜头：完整质量输出 */
    qualityMode?: boolean;
  }): Promise<StructuredInsight> {
    const imageB64 = input.imageBytes.toString("base64");
    const systemPrompt = AGENT_PROMPTS[input.agentId];
    const raw = await vlmService.analyzeImage({
      imageB64,
      systemPrompt,
      locale: input.locale ?? "zh-CN",
      imageCaption: input.imageCaption,
      latitude: input.latitude,
      longitude: input.longitude,
      birthday: input.birthday,
      agentId: input.agentId,
      qualityMode: input.qualityMode,
    });
    raw.agent_id = input.agentId;
    if (input.agentId === AgentId.PALM_READER) {
      normalizePalmReadingRaw(raw);
    }
    const parsed = structuredInsightSchema.safeParse(raw);
    if (!parsed.success) {
      throw new Error(
        `洞察结构校验失败: ${parsed.error.issues
          .slice(0, 3)
          .map((i) => `${i.path.join(".")}: ${i.message}`)
          .join("; ")}`,
      );
    }
    // 几何 path 在 pipeline 层注入；此处只做文案兜底
    if (input.agentId === AgentId.PALM_READER) {
      return sanitizePalmInsight(parsed.data);
    }
    return parsed.data;
  }
}

export function parseAgentOverride(
  value?: string | null,
): AgentId | undefined {
  if (!value) return undefined;
  const parsed = agentIdSchema.safeParse(value);
  if (!parsed.success) {
    console.warn(
      `[analyze] unknown agent_override="${value}", falling back to auto route`,
    );
    return undefined;
  }
  return parsed.data;
}

export const sceneRouter = new SceneRouter();
export const insightPlanner = new InsightPlanner();
